File: src/main/java/de/kitsunealex/projectx/block/BlockXycroniumOre.scala
package de.kitsunealex.projectx.block

import cpw.mods.fml.relauncher.{Side, SideOnly}
import de.kitsunealex.projectx.api.client.RenderTypes
import de.kitsunealex.projectx.api.color.EnumXycroniumColor
import net.minecraft.block.material.Material
import net.minecraft.client.renderer.texture.IIconRegister
import net.minecraft.util.IIcon

class BlockXycroniumOre extends BlockBase("xycronium_ore", Material.rock, EnumXycroniumColor.toStringArray) with TAnimationHandler {

  private var texture: Array[IIcon] = _

  setHardness(1.7F)
  setResistance(1.5F)

  @SideOnly(Side.CLIENT)
  override def registerBlockIcons(register : IIconRegister): Unit = {
    texture = new Array[IIcon](6)
    for(i <- 0 until 5) texture(i) = register.registerIcon(s"projectx:$getBlockName/${getBlockName}_${getSubNames(i)}")
    texture(5) = register.registerIcon(s"projectx:$getBlockName/${getBlockName}_effect")
  }

  @SideOnly(Side.CLIENT)
  override def getIcon(meta: Int, side: Int): IIcon = texture(meta)

  @SideOnly(Side.CLIENT)
  override def getAnimationIcon(meta: Int, side: Int): IIcon = texture(5)

  @SideOnly(Side.CLIENT)
  override def getAnimationColor(meta: Int, side: Int): Int = EnumXycroniumColor.VALUES(meta).getColor.rgba()

  @SideOnly(Side.CLIENT)
  override def getAnimationBrightness(meta: Int, side: Int): Int = 0x00F000F0

  @SideOnly(Side.CLIENT)
  override def getRenderType: Int = RenderTypes.FULL_BLOCK_GLOW

}

File: src/main/java/de/kitsunealex/projectx/block/BlockXycroniumOre.java
/*
 * This file is part of ProjectX.
 * Copyright (c) 2015 - 2018, KitsuneAlex, All rights reserved.
 *
 * ProjectX is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ProjectX is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ProjectX.  If not, see <http://www.gnu.org/licenses/lgpl>.
 */

package de.kitsunealex.projectx.block;

import de.kitsunealex.projectx.init.ModItems;
import de.kitsunealex.projectx.util.Constants;
import de.kitsunealex.projectx.util.EnumXycroniumColor;
import de.kitsunealex.projectx.util.ISubtypeHolder;
import net.minecraft.block.material.Material;
import net.minecraft.block.state.IBlockState;
import net.minecraft.client.renderer.texture.TextureAtlasSprite;
import net.minecraft.client.renderer.texture.TextureMap;
import net.minecraft.item.ItemStack;
import net.minecraft.util.NonNullList;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IBlockAccess;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import java.util.Arrays;

public class BlockXycroniumOre extends BlockAnimationHandler implements ISubtypeHolder {

    @SideOnly(Side.CLIENT)
    private TextureAtlasSprite[] texture;

    public BlockXycroniumOre() {
        super("xycronium_ore", Material.ROCK);
        setHardness(1.6F);
        setResistance(1.5F);
    }

    @Override
    public String[] getSubNames() {
        return Arrays.stream(EnumXycroniumColor.values()).map(EnumXycroniumColor::getName).toArray(String[]::new);
    }

    @Override
    public void getDrops(NonNullList<ItemStack> drops, IBlockAccess world, BlockPos pos, IBlockState state, int fortune) {
        drops.add(new ItemStack(ModItems.XYCRONIUM_CRYSTAL, 4 + fortune, getMetaFromState(state)));
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void registerIcons(TextureMap map) {
        texture = new TextureAtlasSprite[getSubNames().length + 1];

        for(int i = 0; i < texture.length - 1; i++) {
            String path = String.format("blocks/%s/%s_%s", blockName, blockName, getSubNames()[i]);
            texture[i] = map.registerSprite(new ResourceLocation(Constants.MODID, path));
        }

        String path = String.format("blocks/%s/%s_effect", blockName, blockName);
        texture[getSubNames().length] = map.registerSprite(new ResourceLocation(Constants.MODID, path));
    }

    @Override
    @SideOnly(Side.CLIENT)
    public TextureAtlasSprite getTexture(int meta, int side) {
        return texture[meta];
    }

    @Override
    @SideOnly(Side.CLIENT)
    public TextureAtlasSprite getAnimationTexture(int meta, int side) {
        return texture[getSubNames().length];
    }

    @Override
    @SideOnly(Side.CLIENT)
    public int getAnimationColor(int meta, int side) {
        return EnumXycroniumColor.values()[meta].getColorRGBA();
    }

    @Override
    @SideOnly(Side.CLIENT)
    public int getAnimationBrightness(int meta, int side) {
        return 0x00F000F0;
    }

}
